### PR TITLE
Allow named Arguments to be passed to Dto

### PR DESCRIPTION
**Pull request by eltharin**
Allow to change argument order or use variadic argument in dto constructor.

When Dto implements (empty) interface WithNamedArguments, ObjectHydrator will pass an associated key<>value argument to Dto constructor.
Dto can use named arguments or a variadic argument to put values in their properties.

Final Goal (to do after) is to allow `select NEW ItemDto(item.*) FROM App\Entity\Item as item`

updated with with the nested New operator PR.

### CODE REVIEW

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 1]
```diff
@@ -608,6 +608,38 @@ You can also nest several DTO :
 	
 Note that you can only pass scalar expressions or other Data Transfer Objects to the constructor.
 
+If you use your data-transfer objects for multiple query but not ever the same parameters, you can use named arguments or variadic arguments, add ``WithNamedArguments`` to your Dto :
```
Comment (greg0ire) [inline docs/en/reference/dql-doctrine-query-language.rst]:
```suggestion
If you use your data-transfer objects for multiple queries but not necessarily with the same parameters,
you can use named arguments or variadic arguments, add ``WithNamedArguments`` to your Dto :
```

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 2]
```diff
@@ -608,6 +608,38 @@ You can also nest several DTO :
...
+
+.. code-block:: php
+
+    <?php
+
+    use Doctrine\ORM\WithNamedArguments;
+
+    class CustomerDTO implements WithNamedArguments
+    {
+        public function __construct(public string|null $name = null, public string|null $email = null, public string|null $city = null, public mixed|null $value = null)
```
Comment (greg0ire) [inline docs/en/reference/dql-doctrine-query-language.rst]: Let's use linebreaks.

```suggestion
        public function __construct(
                public string|null $name = null,
                public string|null $email = null,
                public string|null $city = null,
                public mixed|null $value = null,
        )
```

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 3]
```diff
@@ -608,6 +608,38 @@ You can also nest several DTO :
...
+        {
+        }
+    }
+
+And then you can select the fields you want in order you want :
```
Comment (greg0ire) [inline docs/en/reference/dql-doctrine-query-language.rst]:
```suggestion
And then you can select the fields you want in the order you want :
```

File: phpcs.xml.dist [hunk 4]
```diff
@@ -273,4 +273,9 @@
         <!-- https://github.com/doctrine/orm/issues/8537 -->
         <exclude-pattern>src/QueryBuilder.php</exclude-pattern>
     </rule>
+
+     <rule ref="Generic.Files.LineEndings.InvalidEOLChar">
+      <exclude name="Generic.Files.LineEndings.InvalidEOLChar"/>
+     </rule>
```
Comment (greg0ire) [inline phpcs.xml.dist]: Why is this necessary?

File: src/Internal/Hydration/AbstractHydrator.php [hunk 5]
```diff
@@ -254,14 +256,15 @@ abstract protected function hydrateAllData(): mixed;
      *                   newObjects?: array<array-key, array{
      *                       class: ReflectionClass,
      *                       args: array,
+     *                       argNames?: array,
```
Comment (greg0ire) [inline src/Internal/Hydration/AbstractHydrator.php]: Can you do better than just `array` here?

File: src/Internal/Hydration/AbstractHydrator.php [hunk 6]
```diff
@@ -341,33 +342,54 @@ protected function gatherRowData(array $data, array &$id, array &$nonemptyCompon
...
-            unset($rowData['newObjects'][$objIndex]);
+            $newObject = $rowData['newObjects'][$ownerIndex . ':' . $argIndex];
+            unset($rowData['newObjects'][$ownerIndex . ':' . $argIndex]);
 
-            $class = $newObject['class'];
-            $args  = $newObject['args'];
-            $obj   = $class->newInstanceArgs($args);
+            $class          = $newObject['class'];
+            $args           = $newObject['args'];
+            $argColumnNames = $newObject['argColumnNames'];
```
Comment (greg0ire) [inline src/Internal/Hydration/AbstractHydrator.php]: I would avoid creating 3 new variables, here and below, but maybe I'm missing something?

File: src/Query/ResultSetMapping.php [hunk 7]
```diff
@@ -552,25 +551,4 @@ public function addMetaResult(
...
-            return $this;
-        }
-
-        $this->nestedNewObjectArguments = array_merge(
-            [$alias => $owner],
-            $this->nestedNewObjectArguments,
-        );
-
-        return $this;
-    }
```
Comment (greg0ire) [inline src/Query/ResultSetMapping.php:575]: This looks a lot like a breaking change.

File: tests/Tests/Models/CMS/CmsAddressDTONamedArgs.php [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+declare(strict_types=1);
+
+namespace Doctrine\Tests\Models\CMS;
+
+use Doctrine\ORM\WithNamedArguments;
+
+class CmsAddressDTONamedArgs implements WithNamedArguments
+{
+    public function __construct(public string|null $country = null, public string|null $city = null, public string|null $zip = null, public CmsAddressDTO|string|null $address = null)
```
Comment (greg0ire) [inline tests/Tests/Models/CMS/CmsAddressDTONamedArgs.php]: linebreaks please

File: tests/Tests/Models/CMS/CmsUserDTONamedArgs.php [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+declare(strict_types=1);
+
+namespace Doctrine\Tests\Models\CMS;
+
+use Doctrine\ORM\WithNamedArguments;
+
+class CmsUserDTONamedArgs implements WithNamedArguments
+{
+    public function __construct(public string|null $name = null, public string|null $email = null, public string|null $address = null, public int|null $phonenumbers = null, public CmsAddressDTO|null $addressDto = null, public CmsAddressDTONamedArgs|null $addressDtoNamedArgs = null)
```
Comment (greg0ire) [inline tests/Tests/Models/CMS/CmsUserDTONamedArgs.php]: Here as well.

File: tests/Tests/ORM/Functional/NewOperatorTest.php [hunk 10]
```diff
@@ -1080,6 +1083,242 @@ public function testShouldSupportNestedNewOperators(): void
...
     }
+
+    public function testNamedArguments(): void
+    {
+        $dql = '
+            SELECT
+                new CmsUserDTONamedArgs(
+                    e.email,
+                    u.name,
+                    CONCAT(a.country, \' \', a.city, \' \', a.zip) AS address
```
Comment (greg0ire) [inline tests/Tests/ORM/Functional/NewOperatorTest.php]: Please use [nowdoc](https://www.php.net/manual/en/language.types.string.php#language.types.string.syntax.nowdoc) so as to avoid this escaping.

File: tests/Tests/ORM/Functional/NewOperatorTest.php [hunk 11]
```diff
@@ -1080,6 +1083,242 @@ public function testShouldSupportNestedNewOperators(): void
...
+        self::assertSame($this->fixtures[1]->username, $result[1]['cmsUserUsername']);
+        self::assertSame($this->fixtures[2]->username, $result[2]['cmsUserUsername']);
+    }
+
+    public function testVariadicArgument(): void
+    {
+        $dql = '
+            SELECT
+                new CmsUserDTOVariadicArg(
+                    CONCAT(a.country, \' \', a.city, \' \', a.zip) AS address,
```
Comment (greg0ire) [inline tests/Tests/ORM/Functional/NewOperatorTest.php]: Here too please

File: tests/Tests/ORM/Functional/NewOperatorTest.php [hunk 12]
```diff
@@ -1080,6 +1083,242 @@ public function testShouldSupportNestedNewOperators(): void
...
+        self::assertSame($this->fixtures[1]->name, $result[1]['user']->name);
+        self::assertSame($this->fixtures[2]->name, $result[2]['user']->name);
+
+        self::assertSame($this->fixtures[0]->email->email, $result[0]['user']->email);
+        self::assertSame($this->fixtures[1]->email->email, $result[1]['user']->email);
+        self::assertSame($this->fixtures[2]->email->email, $result[2]['user']->email);
+
+        self::assertSame($this->fixtures[0]->address->country . ' ' . $this->fixtures[0]->address->city . ' ' . $this->fixtures[0]->address->zip, $result[0]['user']->address);
+        self::assertSame($this->fixtures[1]->address->country . ' ' . $this->fixtures[1]->address->city . ' ' . $this->fixtures[1]->address->zip, $result[1]['user']->address);
+        self::assertSame($this->fixtures[2]->address->country . ' ' . $this->fixtures[2]->address->city . ' ' . $this->fixtures[2]->address->zip, $result[2]['user']->address);
```
Comment (greg0ire) [inline tests/Tests/ORM/Functional/NewOperatorTest.php]: Linebreaks please, and please use `sprintf()`

File: tests/Tests/ORM/Functional/NewOperatorTest.php [hunk 13]
```diff
@@ -1080,6 +1083,242 @@ public function testShouldSupportNestedNewOperators(): void
...
```
Comment (greg0ire) [inline tests/Tests/ORM/Functional/NewOperatorTest.php]: Here as well please.

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 14]
```diff
@@ -608,6 +608,38 @@ You can also nest several DTO :
...
+
+.. code-block:: php
+
+    <?php
+    $query = $em->createQuery('SELECT NEW CustomerDTO(a.city, c.name) FROM Customer c JOIN c.address a');
```
Comment (greg0ire) [inline docs/en/reference/dql-doctrine-query-language.rst]: Wait so what happens here? Which fields of the DTO are supposed to get hydrated? Hydrating `CustomerDTO::$name` with a city feels wrong :thinking:

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 15]
```diff
@@ -608,6 +608,38 @@ You can also nest several DTO :
...
+    $users = $query->getResult(); // array of CustomerDTO
+
+you can either aliases column :
```
Comment (greg0ire) [inline docs/en/reference/dql-doctrine-query-language.rst]: What do you mean here? I would expect an "or" close in your sentence, like "You can either do this or do this".

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 16]
```diff
@@ -608,6 +608,38 @@ You can also nest several DTO :
...
+
+.. code-block:: php
+
+    <?php
+    $query = $em->createQuery('SELECT NEW CustomerDTO(c.name, CONCAT(a.city, ' ' , a.zip) AS value) FROM Customer c JOIN c.address a');
```
Comment (greg0ire) [inline docs/en/reference/dql-doctrine-query-language.rst]: Did you consider a syntax closer to PHP, like `value: CONCAT(a.city, ' ', a.zip)`?

File: src/WithNamedArguments.php [hunk 17]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\ORM;
+
+interface WithNamedArguments
```
Comment (derrabus) [inline src/WithNamedArguments.php]: Can we solve this issue without introducing empty marker interfaces?

**Review by derrabus — DISMISSED** [review]

File: phpcs.xml.dist (hunk 4 above)
Comment (eltharin) [reply phpcs.xml.dist]: my bad, I'm on windows and when I make local test it is

File: docs/en/reference/dql-doctrine-query-language.rst (hunk 14 above)
Comment (eltharin) [reply docs/en/reference/dql-doctrine-query-language.rst]: No, it's exactly the case I want, As the args are named, it's city and name from the Dto witch are filled .
Maybe I have to print the Dto Output for better understand

File: src/WithNamedArguments.php (hunk 17 above)
Comment (eltharin) [reply src/WithNamedArguments.php]: We can have `$em->createQuery('SELECT NEW NAMED CustomerDTO(...);`

But I find it less readable

File: docs/en/reference/dql-doctrine-query-language.rst (hunk 16 above)
Comment (eltharin) [reply docs/en/reference/dql-doctrine-query-language.rst]: I can think about it but since we are in a SQL/DQL query I stayed in this way

File: src/Query/ResultSetMapping.php (hunk 7 above)
Comment (eltharin) [reply src/Query/ResultSetMapping.php:575]: This function was created in my previous PR, so not very break

File: docs/en/reference/dql-doctrine-query-language.rst (hunk 14 above)
Comment (greg0ire) [reply docs/en/reference/dql-doctrine-query-language.rst]: Yes, and also, you could explain what is going to happen: `And then you can select the fields you want in the order you want, and the ORM will try to match argument names with the selected field names`. If you do that, you should probably explain what happens if one of the selected field matches, but not the other. Or what happens if we do `SELECT NEW CustomerDTO(a.city, b.city) FROM …`

File: src/WithNamedArguments.php (hunk 17 above)
Comment (eltharin) [reply src/WithNamedArguments.php]: Do you prefer have a const value USE_NAMED_ARGS in dto instead of implents empty interface  ?

File: docs/en/reference/dql-doctrine-query-language.rst (hunk 16 above)
Comment (eltharin) [reply docs/en/reference/dql-doctrine-query-language.rst]: It works, but I really hate this notation in an SQL query, so I left AS keywork working too.

File: src/WithNamedArguments.php (hunk 17 above)
Comment (derrabus) [reply src/WithNamedArguments.php]: The gods of PHP gave us attributes, so we don't need marker interfaces anymore. If you really really need to flag the DTO class in some way, please use an attribute for that.

That being said, I think I'd prefer a dedicated DQL syntax for named argument calls. That should make any marking of DTO classes obsolete.

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 18]
```diff
@@ -608,6 +608,60 @@ You can also nest several DTO :
...
+If you use your data-transfer objects for multiple queries but not necessarily with the same parameters,
+you can use named arguments or variadic arguments, add ``WithNamedArguments`` to your Dto :
+
+.. code-block:: php
+
+    <?php
+
+    use Doctrine\ORM\WithNamedArguments;
+
+    class CustomerDTO implements WithNamedArguments
```
Comment (derrabus) [inline docs/en/reference/dql-doctrine-query-language.rst]: Now that you've removed that interface, the documentation needs to be adjusted.

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 19]
```diff
@@ -608,6 +608,60 @@ You can also nest several DTO :
...
+    {
+        public function __construct(
+                public string|null $name = null,
+                public string|null $email = null,
+                public string|null $city = null,
+                public mixed|null $value = null,
+        )
+        {
```
Comment (derrabus) [inline docs/en/reference/dql-doctrine-query-language.rst]:
```suggestion
        ) {
```

File: src/Query/ResultSetMapping.php (hunk 7 above)
Comment (derrabus) [reply src/Query/ResultSetMapping.php:575]: Indeed, we haven't issued a release yet that includes this method. The change is fine.

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 20]
```diff
@@ -608,6 +608,68 @@ You can also nest several DTO :
...
+
+ORM will also look column aliases before columns names :
+
+.. code-block:: php
+
+    <?php
+    $query = $em->createQuery('SELECT NEW NAMED CustomerDTO(c.name, CONCAT(a.city, ' ' , a.zip) AS value) FROM Customer c JOIN c.address a');
+    $users = $query->getResult(); // array of CustomerDTO
+
+	// CustomerDTO => {name : 'DOE', email: null, city: null, value: 'New York 10011'}
```
Comment (greg0ire) [inline docs/en/reference/dql-doctrine-query-language.rst]: You used a tab here. Please check the whole diff and make sure you never use tabs.

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 21]
```diff
@@ -608,6 +608,68 @@ You can also nest several DTO :
...
+
+.. code-block:: php
+
+	<?php
+	$query = $em->createQuery('SELECT NEW NAMED CustomerDTO(name: c.name, value: CONCAT(a.city, ' ' , a.zip)) FROM Customer c JOIN c.address a');
+	$users = $query->getResult(); // array of CustomerDTO
+
+	// CustomerDTO => {name : 'DOE', email: null, city: null, value: 'New York 10011'}
+
+The `NAMED` keywork must be added to all dto you want to instantiate :
```
Comment (greg0ire) [inline docs/en/reference/dql-doctrine-query-language.rst]:
```suggestion
The ``NAMED`` keyword must be added to all dto you want to instantiate :
```

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 22]
```diff
@@ -608,6 +608,68 @@ You can also nest several DTO :
...
+
+.. code-block:: php
+
+	<?php
+	$query = $em->createQuery('SELECT NEW NAMED CustomerDTO(name: c.name, address: NEW NAMED AddressDTO(a.street, a.city, a.zip)) FROM Customer c JOIN c.address a');
+	$users = $query->getResult(); // array of CustomerDTO
+
+	// CustomerDTO => {name : 'DOE', email: null, city: null, value: 'New York 10011'}
+
+Be careful, if two fields have the same name/alias, second will erase first.
```
Comment (greg0ire) [inline docs/en/reference/dql-doctrine-query-language.rst]:
```suggestion
Be careful, if two fields have the same name/alias, the second one will overwrite the first one.
```

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 23]
```diff
@@ -608,6 +608,68 @@ You can also nest several DTO :
...
+If you put an argument name empty (for functions) or not in constructor argument list, ORM will throw an error.
```
Comment (greg0ire) [inline docs/en/reference/dql-doctrine-query-language.rst]: > If you put an argument name empty (for functions)

What do you mean by that?

File: docs/en/reference/dql-doctrine-query-language.rst (hunk 22 above)
Comment (greg0ire) [inline docs/en/reference/dql-doctrine-query-language.rst]: Should it instead throw an exception?

File: docs/en/reference/dql-doctrine-query-language.rst (hunk 23 above)
Comment (greg0ire) [inline docs/en/reference/dql-doctrine-query-language.rst]: An error or an exception?

**Comment by greg0ire** [comment]
Please kindly squash your commits together. If you don't, we'll try to remember to do it for you but it's best if you save us this trouble.

How to do that?

1. `git rebase -i origin/3.3.x`, assuming `origin` is a git remote that points to this repository, and not your fork. If you're not sure what your remotes are, run `git remote -vvv`, there should be your fork and the holy/reference/base/origin/whatever-you-call-it repository.
2. A window will show up with many lines, replace `pick` with `fixup` on every line but the first one
3. Close your editor, git should do its magic, and you should end up with one commit
4. Use `git push --force` to overwrite what you already push. Don't forget the `--force` option otherwise git will try to merge both things together.

File: docs/en/reference/dql-doctrine-query-language.rst (hunk 23 above)
Comment (eltharin) [reply docs/en/reference/dql-doctrine-query-language.rst]: actually an error, thorwn by PHP and ORM : "Error: Unknown named parameter $"

Comment (greg0ire) [reply docs/en/reference/dql-doctrine-query-language.rst:671]: Sorry, where in the code is it thrown? I can't find it.

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 24]
```diff
@@ -591,23 +591,85 @@ You can also nest several DTO :
...
+
+.. code-block:: php
+
+    <?php
+    $query = $em->createQuery('SELECT NEW NAMED CustomerDTO(name: c.name, address: NEW NAMED AddressDTO(a.street, a.city, a.zip)) FROM Customer c JOIN c.address a');
+    $users = $query->getResult(); // array of CustomerDTO
+
+    // CustomerDTO => {name : 'DOE', email: null, city: null, value: 'New York 10011'}
+
+If two fields have the same name/alias, a RepeatedAliasException is thrown.
```
Comment (greg0ire) [inline docs/en/reference/dql-doctrine-query-language.rst]:
```suggestion
If two fields have the same name/alias, a ``RepeatedAliasException`` is thrown.
```

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 25]
```diff
@@ -591,23 +591,85 @@ You can also nest several DTO :
...
+If a field without column name (functions for example) has no alias set, a MissingAliasException is thrown.
```
Comment (greg0ire) [inline docs/en/reference/dql-doctrine-query-language.rst]:
```suggestion
If a field without a column name (when using functions for example) has no alias set, a ``MissingAliasException`` is thrown.
```

Comment (greg0ire) [reply docs/en/reference/dql-doctrine-query-language.rst]: [When I use a function](https://sqlfiddle.com/postgresql/online-compiler?id=26728015-ebc0-4603-8e34-69c6cc60fa4a) the resulting column does have a name :thinking:

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 26]
```diff
@@ -591,23 +591,85 @@ You can also nest several DTO :
...
+
+.. code-block:: php
+
+    <?php
+    $query = $em->createQuery('SELECT NEW NAMED CustomerDTO(c.name, CONCAT(a.city, ' ' , a.zip) AS value) FROM Customer c JOIN c.address a');
+    $users = $query->getResult(); // array of CustomerDTO
+
+    // CustomerDTO => {name : 'DOE', email: null, city: null, value: 'New York 10011'}
+
+For aliases a field you can use SQL notation with `AS` keyword, or the PHP named arguments notation :
```
Comment (greg0ire) [inline docs/en/reference/dql-doctrine-query-language.rst]:
```suggestion
In order to  alias a field you can either use the SQL keyword ``AS``, or PHP's named arguments syntax :
```

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 27]
```diff
@@ -591,23 +591,85 @@ You can also nest several DTO :
...
+
+.. code-block:: php
+
+    <?php
+    $query = $em->createQuery('SELECT NEW NAMED CustomerDTO(name: c.name, value: CONCAT(a.city, ' ' , a.zip)) FROM Customer c JOIN c.address a');
```
Comment (greg0ire) [inline docs/en/reference/dql-doctrine-query-language.rst]: The example only illustrates the second method. Also, you are saying "For aliases a field", but with the second method, there is no aliasing taking place at all, right?

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 28]
```diff
@@ -591,23 +591,85 @@ You can also nest several DTO :
...
+    $users = $query->getResult(); // array of CustomerDTO
+
+    // CustomerDTO => {name : 'DOE', email: null, city: null, value: 'New York 10011'}
+
+The ``NAMED`` keyword must be added to all dto you want to instantiate :
```
Comment (greg0ire) [inline docs/en/reference/dql-doctrine-query-language.rst]:
```suggestion
The ``NAMED`` keyword must precede all DTO you want to instantiate :
```

File: docs/en/reference/dql-doctrine-query-language.rst (hunk 27 above)
Comment (eltharin) [reply docs/en/reference/dql-doctrine-query-language.rst]: It's for showing we can make a mix of field names and aliases

Comment (greg0ire) [reply docs/en/reference/dql-doctrine-query-language.rst]: In the example though, it's 100% field names, right? Or is there an alias I am missing?

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 29]
```diff
@@ -591,23 +591,85 @@ You can also nest several DTO :
...
 
     <?php
     $query = $em->createQuery('SELECT NEW CustomerDTO(c.name, e.email, NEW AddressDTO(a.street, a.city, a.zip)) FROM Customer c JOIN c.email e JOIN c.address a');
     $users = $query->getResult(); // array of CustomerDTO
-	
+
 Note that you can only pass scalar expressions or other Data Transfer Objects to the constructor.
 
+If you use your data-transfer objects for multiple queries but not necessarily with the same parameters,
+you can use named arguments or variadic arguments, add ``named`` keyword in your DQL querie before your Dto :
```
Comment (greg0ire) [inline docs/en/reference/dql-doctrine-query-language.rst]:
```suggestion
you can use named arguments or variadic arguments, add the ``named`` keyword in your DQL query before your DTO :
```

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 30]
```diff
@@ -591,23 +591,85 @@ You can also nest several DTO :
...
+    <?php
+
+    class CustomerDTO
+    {
+        public function __construct(
+                public string|null $name = null,
+                public string|null $email = null,
+                public string|null $city = null,
+                public mixed|null $value = null,
+                public AddressDTO|null $address = null,
```
Comment (greg0ire) [inline docs/en/reference/dql-doctrine-query-language.rst]:
```suggestion
            public string|null $name = null,
            public string|null $email = null,
            public string|null $city = null,
            public mixed|null $value = null,
            public AddressDTO|null $address = null,
```

File: src/Exception/MissingAliasException.php [hunk 31]
```diff
@@ -0,0 +1,15 @@
...
+
+namespace Doctrine\ORM\Exception;
+
+use LogicException;
+
+class MissingAliasException extends LogicException implements ORMException
+{
+    public static function create(): self
+    {
+        return new self('Alias is missing on a select part.');
```
Comment (greg0ire) [inline src/Exception/MissingAliasException.php]: Would it be possible to get more details here? Like which select part is missing an alias?

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 32]
```diff
@@ -591,23 +591,85 @@ You can also nest several DTO :
...
 
     <?php
     $query = $em->createQuery('SELECT NEW CustomerDTO(c.name, e.email, NEW AddressDTO(a.street, a.city, a.zip)) FROM Customer c JOIN c.email e JOIN c.address a');
     $users = $query->getResult(); // array of CustomerDTO
-	
+
 Note that you can only pass scalar expressions or other Data Transfer Objects to the constructor.
 
+If you use your data-transfer objects for multiple queries but not necessarily with the same parameters,
+you can use named arguments or variadic arguments, add the ``named`` keyword in your DQL query before your DTO :
```
Comment (greg0ire) [inline docs/en/reference/dql-doctrine-query-language.rst]: Here you mentioned named arguments and variadic arguments. Is it necessary to add the `NAMED` keyword to use variadic arguments?

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 33]
```diff
@@ -591,23 +591,85 @@ You can also nest several DTO :
...
 .. code-block:: php
 
     <?php
     $query = $em->createQuery('SELECT NEW CustomerDTO(c.name, e.email, NEW AddressDTO(a.street, a.city, a.zip)) FROM Customer c JOIN c.email e JOIN c.address a');
     $users = $query->getResult(); // array of CustomerDTO
-	
+
 Note that you can only pass scalar expressions or other Data Transfer Objects to the constructor.
 
+If you use your data-transfer objects for multiple queries but not necessarily with the same parameters,
```
Comment (greg0ire) [inline docs/en/reference/dql-doctrine-query-language.rst]:
```suggestion
If you use your data transfer objects for multiple queries but not necessarily with the same parameters,
```

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 34]
```diff
@@ -591,23 +591,85 @@ You can also nest several DTO :
...
+you can use named arguments or variadic arguments, add the ``named`` keyword in your DQL query before your DTO :
+
+.. code-block:: php
```
Comment (greg0ire) [inline docs/en/reference/dql-doctrine-query-language.rst:616]: Because of what precedes, I'm expecting to see a DQL query. I think you should reorganize this.
If you use your data transfer objects for multiple queries, and you would rather not have to
specify arguments that precede the ones you are really interested in, you can use named arguments.

Consider the following DTO, which uses optional arguments:

// code for CustomerDTO here

You can specify arbitrary arguments in an arbitrary order by using the named argument syntax, and the ORM will try to match argument names with the selected column names.
The syntax relies on the `NAMED` keyword, like so:

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 35]
```diff
@@ -591,23 +591,85 @@ You can also nest several DTO :
...
+
+    <?php
+    $query = $em->createQuery('SELECT NEW NAMED CustomerDTO(a.city, c.name) FROM Customer c JOIN c.address a');
+    $users = $query->getResult(); // array of CustomerDTO
+
+    // CustomerDTO => {name : 'SMITH', email: null, city: 'London', value: null}
+
+ORM will also look column aliases before columns names :
```
Comment (greg0ire) [inline docs/en/reference/dql-doctrine-query-language.rst]:
```suggestion
ORM will also give precedence to column aliases over column names :
```

File: src/Exception/DuplicateFieldException.php [hunk 36]
```diff
@@ -0,0 +1,17 @@
...
+
+use LogicException;
+
+use function sprintf;
+
+class DuplicateFieldException extends LogicException implements ORMException
+{
+    public static function create(string $argName, string $columnName): self
+    {
+        return new self(sprintf('Name %s for `%s` already in use.', $argName, $columnName));
```
Comment (greg0ire) [inline src/Exception/DuplicateFieldException.php]:
```suggestion
        return new self(sprintf('Name "%s" for "%s" already in use.', $argName, $columnName));
```

File: src/Exception/NoMatchingPropertyException.php [hunk 37]
```diff
@@ -0,0 +1,17 @@
...
+
+use LogicException;
+
+use function sprintf;
+
+class NoMatchingPropertyException extends LogicException implements ORMException
+{
+    public static function create(string $property): self
+    {
+        return new self(sprintf('Column Name `%s` has no property name or alias.', $property));
```
Comment (greg0ire) [inline src/Exception/NoMatchingPropertyException.php]:
```suggestion
        return new self(sprintf('Column name "%s" does not match any property name. Consider aliasing it to the name of an existing property.', $property));
```

File: src/Query/Parser.php [hunk 38]
```diff
@@ -1764,29 +1774,71 @@ public function NewObjectExpression(): AST\NewObjectExpression
...
+
+        if ($useNamedArguments) {
+            $startToken = $this->lexer->lookahead?->position ?? 0;
+
+            $newArg = $this->NewObjectArg($fieldAlias);
+
+            $key = $fieldAlias ?? $newArg->field ?? null;
+
+            if ($key === null) {
+                throw NoMatchingPropertyException::create(trim(substr(($this->query->getDQL() ?? ''), $startToken, ($this->lexer->lookahead->position ?? 0) - $startToken)));
```
Comment (greg0ire) [inline src/Query/Parser.php]: Let's use linebreaks please.

```suggestion
                throw NoMatchingPropertyException::create(trim(substr(
                    ($this->query->getDQL() ?? ''),
                    $startToken,
                    ($this->lexer->lookahead->position ?? 0) - $startToken,
                )));
```

File: src/Query/Parser.php [hunk 39]
```diff
@@ -1764,29 +1774,71 @@ public function NewObjectExpression(): AST\NewObjectExpression
...
+            }
+
+            if (array_key_exists($key, $args)) {
+                throw DuplicateFieldException::create($key, trim(substr(($this->query->getDQL() ?? ''), $startToken, ($this->lexer->lookahead->position ?? 0) - $startToken)));
```
Comment (greg0ire) [inline src/Query/Parser.php]:
```suggestion
                throw DuplicateFieldException::create($key, trim(substr(
                    ($this->query->getDQL() ?? ''),
                    $startToken,
                    ($this->lexer->lookahead->position ?? 0) - $startToken
                )));
```

File: docs/en/reference/dql-doctrine-query-language.rst (hunk 32 above)
Comment (eltharin) [reply docs/en/reference/dql-doctrine-query-language.rst]: variadic argument can be used without ``named`` keyword but the array will be "indexed" while with ``named``it will be associated

**Review by greg0ire — DISMISSED** [review]

**Review by SenseException — DISMISSED** [review]

**Comment by greg0ire** [comment]
@derrabus please give this another review.

**Comment by eltharin** [comment]
waiting for https://github.com/doctrine/orm/pull/11619 to push EBNF update

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 40]
```diff
@@ -591,23 +591,88 @@ You can also nest several DTO :
...
+    $users = $query->getResult(); // array of CustomerDTO
+
+    // CustomerDTO => {name : 'DOE', email: null, city: null, value: 'New York 10011'}
+
+To define a custom name for a DTO constructor argument, you can either alias the column with the ``AS`` keyword, or with PHP's named arguments syntax.
+
+.. code-block:: php
+
+    <?php
+    $query = $em->createQuery('SELECT NEW NAMED CustomerDTO(name: c.name, value: CONCAT(a.city, ' ' , a.zip)) FROM Customer c JOIN c.address a');
```
Comment (beberlei) [inline docs/en/reference/dql-doctrine-query-language.rst]: I would prefer if we have only this syntax available `CustomerDTO(name: c.name, value: CONCAT(a.city, ' ' , a.zip))` and not the two above that implicitly map the name or the AS naming.

This would in my opinion also make the `NAMED` keyword obsolete, which I would like to get rid of as well.

The goal should be that named arguments syntax looks exactly the same as in PHP itself, because the `NEW className` somewhat implies that this is what DQL supports.

Comment (eltharin) [reply docs/en/reference/dql-doctrine-query-language.rst]: Disable implicity map means that we have to write all the field's names for our queries, instead of add a keyword "named".

Remove AS keyword means for those who actually have a named array as result (with AS keyword) to rewrite all the names while in my case just have to add NEW NAMED myDTO(...) and it's done.

Forcing to write: name: c.name, lastname: c.lastname, age: c.age, birthdate: c.birthdate, ... is in my opinion a waste of time for the few fields to rename.

For those as me who have helpers to select fields to add in queries, remove NAMED keyword means to be obliged to name ALL the fields.

As I said in descritpion, final goal is to be able to write c.* and have all the fields by the names, 

That's why the first version of this PR have only AS keyword and not the 'PHP syntax'

Comment (beberlei) [reply docs/en/reference/dql-doctrine-query-language.rst]: @eltharin I discussed with @greg0ire and @derrabus at our Doctrine core team meetup in person, and I guess the main thing is that we want to avoid adding 2 syntaxes (implicit and AS naming, plus name double colon). We considered both and your arguments and then we would rather remove the PHP-style named arguments syntax and only keep the examples 1+2, so the following should be allowed:

* SELECT NEW NAMED CustomerDTO(a.city, c.name) FROM Customer c JOIN c.address a')
* SELECT NEW NAMED CustomerDTO(c.name, CONCAT(a.city, ' ' , a.zip) AS value) FROM Customer c JOIN c.address a

and this syntax should be removed again:

* SELECT NEW NAMED CustomerDTO(name: c.name, value: CONCAT(a.city, ' ' , a.zip)) FROM Customer c JOIN c.address a

Everything else is 👍 from me.

Comment (eltharin) [reply docs/en/reference/dql-doctrine-query-language.rst]: thank you to all three of you  (and other mainteners) for being so open-minded and for your work.

**Review by greg0ire — APPROVED** [review]

**Comment by greg0ire** [comment]
Thanks @eltharin !

**Comment by beberlei** [comment]
@eltharin 💯 thanks
